google-cloud-spanner/lib/google/cloud/spanner/fields.rb

```diff
@@ -41,20 +41,75 @@ module Spanner
       #
       class Fields
         ##
-        # @private
+        # Creates {Fields} object from types. See {Client#fields}.
+        #
+        # This object can be used to create {Data} objects by providing values
+        # that match the field types. See {Fields#struct}.
+        #
+        # @param [Array, Hash] types Accepts an array or hash types.
+        #
+        #   Arrays can contain just the type value, or a sub-array of the
+        #   field's name and type value. Hash keys must contain the field name
+        #   as a `Symbol` or `String`, or the field position as an `Integer`.
+        #   Hash values must contain the type value. If a Hash is used the
+        #   fields will be created using the same order as the Hash keys.
+        #
+        #   Supported type values incude:
+        #
+        #   * `:BOOL`
+        #   * `:BYTES`
+        #   * `:DATE`
+        #   * `:FLOAT64`
+        #   * `:INT64`
+        #   * `:STRING`
+        #   * `:TIMESTAMP`
+        #   * `Array` - Lists are specified by providing the type code in an
+        #     array. For example, an array of integers are specified as
+        #     `[:INT64]`.
+        #   * {Fields} - Nested Structs are specified by providing a Fields
+        #     object.
+        #
+        # @return [Fields] The fields of the given types.
+        #
+        # @example Defining Fields/Structs with named fields
+        #   require "google/cloud/spanner"
+        #
+        #   named_type = Google::Cloud::Spanner::Fields.new(
+        #     { id: :INT64, active: :BOOL }
+        #   )
+        #   named_data = named_type.struct({ id: 42, active: false })
+        #
+        # @example Defining Fields/Structs with anonymous fields
+        #   require "google/cloud/spanner"
+        #
+        #   anon_type = Google::Cloud::Spanner::Fields.new(
+        #     [:INT64, :BOOL]
+        #   )
+        #   anon_data = anon_type.struct [42, false]
+        #
+        # @example Defining Fields/Structs with duplicate field names
+        #   require "google/cloud/spanner"
+        #
+        #   dup_type = Google::Cloud::Spanner::Fields.new(
+        #     [[:x, :INT64], [:x, :BOOL]]
+        #   )
+        #   dup_data = dup_type.struct [42, false]
+        #
         def initialize types
-          @fields = []
-          if types.is_a? Array
-            types.each do |type|
-              @fields << field(type)
-            end
-          elsif types.is_a? Hash
-            types.each do |type|
-              @fields << field(type)
-            end
-          else
+          types = types.to_a if types.is_a? Hash
+
+          unless types.is_a? Array
             raise ArgumentError, "can only accept Array or Hash"
           end
+
+          # Verify positional arguments are in order
+          types.each_with_index do |type, index|
+            if type.is_a?(Array) && type.first.is_a?(Integer)
+              raise "types must be in order" if type.first != index
```

Comment:
Should we have custom `Error`s ? instead of raising anonymous/generic errors.

Reply:
Yes, this what I meant :) not just a raise ..

Reply:
This error was a placeholder until support for unsorted positional types was added. This code has been removed.